# Retrieval Augmented Generation (RAG) on Large Language Model (LLM)

Large Language Models (LLMs) have been developing at a remarkable pace in recent years, demonstrating outstanding capabilities in generating human-like text, answering questions, and supporting various natural language processing tasks. Some well-known applications that implement these models include ChatGPT and Gemini.

In this project, I will explore the use of a tool called Retrieval Augmented Generation (RAG), which can improve LLM accuracy by retrieving relevant information. With this approach, the LLM works not only based on its training data but also leverages external data sources that we provide as input.

## Libraries

In [4]:
# Libraries
import numpy as np
import pandas as pd
import faiss
from dotenv import load_dotenv
import os
from openai import OpenAI
from sentence_transformers import SentenceTransformer, util

## Import Dataset

In [5]:
games = pd.read_csv('data/vgsales.csv')
games.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

## Data Description

The data used in this project is [Video Game Sales](https://www.kaggle.com/datasets/gregorut/videogamesales?resource=download), sourced from [Kaggle](https://kaggle.com). It contains a list of video games that have sold more than 100,000 copies. This dataset includes the following features:

- `Rank` : Ranking of overall sales
- `Name` : The games name
- `Platform` : Platform of the games release (i.e. PC,PS4, etc.)
- `Year` : Year of the game's release
- `Genre` : Genre of the game
- `Publisher` : Publisher of the game
- `NA_Sales` : Sales in North America (in millions)
- `EU_Sales` : Sales in Europe (in millions)
- `JP_Sales` : Sales in Japan (in millions)
- `Other_Sales` : Sales in the rest of the world (in millions)
- `Global_Sales` : Total worldwide sales.


## Data Preparation

To ensure proper ordering, the dataset will first be sorted by the ranking column. For this project, the data that will be used is the top 200 entries, as this selective approach helps optimize processing efficiency. This data limitation is implemented because the model to be used later is the free version, as it helps accommodate the platform’s feature restrictions while still providing meaningful results.

In [6]:
# Sort data based on 'Rank'
games.sort_values(by = 'Rank', ascending = True)

# Extract a subset of the top 200 entries
games = games.head(200)

### Missing Values Check

In [7]:
# Check for missing values
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          200 non-null    int64  
 1   Name          200 non-null    object 
 2   Platform      200 non-null    object 
 3   Year          199 non-null    float64
 4   Genre         200 non-null    object 
 5   Publisher     200 non-null    object 
 6   NA_Sales      200 non-null    float64
 7   EU_Sales      200 non-null    float64
 8   JP_Sales      200 non-null    float64
 9   Other_Sales   200 non-null    float64
 10  Global_Sales  200 non-null    float64
dtypes: float64(6), int64(1), object(4)
memory usage: 17.3+ KB


Based on the result, there is a missing value in column `Year`.

In [8]:
games[games.isnull().any(axis=1)]

Rank             Name Platform  Year   Genre        Publisher  NA_Sales  \
179   180  Madden NFL 2004      PS2   NaN  Sports  Electronic Arts      4.26   

     EU_Sales  JP_Sales  Other_Sales  Global_Sales  
179      0.26      0.01         0.71          5.23

To handle these missing values, I will apply imputation method. The data that I will impute is based on the information that I got from [maddenratings.weebly.com](https://maddenratings.weebly.com/madden-nfl-2004.html#:~:text=Release%20date%3A%20August%2012%2C%202003,on%20the%202003%20NFL%20season.)

In [9]:
# Imputation value
games.loc[179, 'Year'] = 2003

# Check for missing values
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          200 non-null    int64  
 1   Name          200 non-null    object 
 2   Platform      200 non-null    object 
 3   Year          200 non-null    float64
 4   Genre         200 non-null    object 
 5   Publisher     200 non-null    object 
 6   NA_Sales      200 non-null    float64
 7   EU_Sales      200 non-null    float64
 8   JP_Sales      200 non-null    float64
 9   Other_Sales   200 non-null    float64
 10  Global_Sales  200 non-null    float64
dtypes: float64(6), int64(1), object(4)
memory usage: 17.3+ KB


Now, the **data is free from missing values**.

### Duplicate Values Check

In [10]:
games.duplicated().sum()

0

There is **no duplicated values in this data**.

### Formatting Data Display

To tidy up the data display so that the model can read it properly, I will make several adjustments, namely:

- Removing the decimals in the `Year` column by converting its data type to integer
- Adding the suffix “Million” to the `_Sales` data columns
- Rename `Name` to `Game_Name` and `_Sales` data columns name from abbreviation to full name

In [11]:
# Change 'Year' data type to integer
games['Year'] = games['Year'].astype('int64')

# Add 'Million' suffix to sales columns
columns_sales = ['NA_Sales', 'EU_Sales', 'JP_Sales', "Other_Sales", 'Global_Sales']

for col in columns_sales:
    games[col] = games[col].astype(str) + ' Million'

# Rename columns
games.rename(columns={'Name': 'Game_Name',
                      'NA_Sales': 'North_America_Sales',
                      'EU_Sales': 'Europe_Sales',
                      'JP_Sales': 'Japan_Sales'}, inplace=True) 

games.head()

Rank                 Game_Name Platform  Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996  Role-Playing  Nintendo   

  North_America_Sales   Europe_Sales    Japan_Sales   Other_Sales  \
0       41.49 Million  29.02 Million   3.77 Million  8.46 Million   
1       29.08 Million   3.58 Million   6.81 Million  0.77 Million   
2       15.85 Million  12.88 Million   3.79 Million  3.31 Million   
3       15.75 Million  11.01 Million   3.28 Million  2.96 Million   
4       11.27 Million   8.89 Million  10.22 Million   1.0 Million   

    Global_Sales  
0  82.74 Million  
1  40.24 Million  
2  35.82 Million  
3   33.0 Million  
4  31.37 Million

In [ ]:
# Export processed dataset to csv for application later
games.to_csv("games_top200.csv", index = False)

### Transform Data to Text Format

Since tabular data is challenging for LLM to process effectively, I will transform each row of data (where values are separated by columns) into a one sentence. This structured textual format will make the data easier for the LLM to learn and interpret. The resulting sentences will be stored in a new column named `Text`.

In [12]:
selected_cols = games.columns.to_list()

games['Text'] = games[selected_cols].apply(
    lambda x: ' | '.join(f'{col} : {x[col]}' for col in selected_cols),
    axis=1
)

print(games['Text'].head())

0    Rank : 1 | Game_Name : Wii Sports | Platform :...
1    Rank : 2 | Game_Name : Super Mario Bros. | Pla...
2    Rank : 3 | Game_Name : Mario Kart Wii | Platfo...
3    Rank : 4 | Game_Name : Wii Sports Resort | Pla...
4    Rank : 5 | Game_Name : Pokemon Red/Pokemon Blu...
Name: Text, dtype: object


## Embedding Text

Text Embedding is the process of converting text into numerical representations (vectors) that capture the meaning and context of the text. These vectors are then used to measure similarity between different pieces of data. In the RAG process, the vectors of our stored data are compared with the vector of a given query/question. This allows the LLM to retrieve and generate the most relevant answers based on semantic similarity.

The embedding model I will use is **paraphrase-MiniLM-L6-v2**. This model is part of the Sentence Transformers library and is built on top of BERT architecture. The text embedding process will use the `encode()` function with the parameter `convert_to_numpy=True`.

In [13]:
# Prepare Embedding Model
model = SentenceTransformer("paraphrase-MiniLM-L6-v2")

To better visualize the text embedding process and vector similarity search, I'll conduct a test query using a simple question. For this initial test, I'll only use the top 5 data entries first.

In [14]:
# Query example test
query_test = ["Which game has the highest global sales?"]

# Data for output example test
output_test = games['Text'].head(5)

# Embedding text
embedding_query_test = model.encode(query_test, convert_to_numpy= True)
embedding_output_test = model.encode(output_test, convert_to_numpy= True)

## Cosine Similarity

This method enables semantic similarity comparison between query data and reference datasets. Here are the scoring rules:

- **Values closer to 1** indicate higher similarity
- **Values closer to 0** indicate lower similarity

The similarity measurement is calculated using the `util.cos_sim()` function from the Sentence Transformers library.

In [15]:
cosine_scores_test = util.cos_sim(embedding_query_test, embedding_output_test)
cosine_scores_test

tensor([[0.6459, 0.5597, 0.5819, 0.6049, 0.5833]])

In [16]:
games['Global_Sales'].head(5)

0    82.74 Million
1    40.24 Million
2    35.82 Million
3     33.0 Million
4    31.37 Million
Name: Global_Sales, dtype: object

With query:

- Which game has the highest global sales?

The highest similarity score (0.6459) corresponds to `index [0]`. Therefore, the system will tend to use the data with `index [0]` as the answer.

## FAISS Indexing

To optimize vector similarity search performance, I will implement FAISS (Facebook AI Similarity Search). FAISS is a highly optimized library developed by Facebook AI Research specifically designed for efficient similarity search in high-dimensional vector spaces.

In RAG systems, FAISS serves as the core retrieval engine to quickly identify the most semantically relevant documents or data points from a large knowledge base. FAISS enables efficient indexing and retrieval of the most semantically similar data points.

Before indexing, I'll manually calculate cosine similarity since there's no built-in function to directly integrate cosine similarity with FAISS. This calculation processes the vector dimensions from the embedding results.


In [17]:
# Calculate cosine similarity
embedding_output_test = embedding_output_test / np.linalg.norm(embedding_output_test, axis=1, keepdims=True)
embedding_output_test = embedding_output_test.astype('float32')

# Collect embedding data dimension
dimension = embedding_output_test.shape[1]
dimension

384

Next, the index can be augmented using the `IndexFlatL2()` function and incorporated with the embeddings via the `.add()` method.

In [18]:
# Indexing vector data
index = faiss.IndexFlatL2(dimension)
index.add(embedding_output_test)

Using FAISS, we can filter search results to the top *k* best matches to reduce computational load during retrieval for a query. To do this, we use the `.search()` function with parameters containing the query and the number of search results `k`. That function resulting:

- `D:` distance between the query vector with nearest neighbors.
- `I:` indeks or positions of the most similar vectors in your dataset.

In [19]:
# Filter the data to top 2 (k=2) best matches
D, I = index.search(embedding_query_test, k = 2)
print(D)
print(I)

[[46.096603 46.702133]]
[[0 3]]


Based on the result, `index [0]` is the nearest neighbors vector for the given query. It is aligned with cosine similarity scores. This validates our FAISS implementation's ability to retrieve contextually relevant results. 

## Custom Function for FAISS - Cosine Indexing

Custom function will be created by combining all previously implemented steps up to the index augmentation stage.

In [20]:
def faiss_cosine_index(data):
    # Embedding data
    embedding = model.encode(data , convert_to_numpy=True)

    # Calculate for cosine similarity
    embedding = embedding / np.linalg.norm(embedding, axis=1, keepdims=True)
    embedding = embedding.astype('float32')

    # FAISS Indexing
    dimension = embedding.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(embedding)

    return index, embedding

Next, the custom `faiss_cosine_index` function will be tested by retrieving the distances and indices of the top 5 most similar data entries (k=5) matching the query.

In [21]:
index, embedding = faiss_cosine_index(games["Text"])

In [22]:
D, I = index.search(embedding_query_test, k = 5)

In [23]:
print(D)
print(I)

[[46.096603 46.375694 46.43065  46.643032 46.675224]]
[[  0  10  13 138  77]]


The results confirm that the custom function `faiss_cosine_index` has been successfully implemented and can generate distances and indices for data matching the query.

Futhermore, a new function will be created to filter the highest-similarity data and present it in a dataframe format.

In [24]:
def retrieve(query, index, df, top_k=3):
    # Embedding and normalize the query
    query_embedding = model.encode([query], convert_to_numpy=True)
    query_embedding = query_embedding / np.linalg.norm(query_embedding, axis=1, keepdims=True)
    query_embedding = query_embedding.astype("float32")

    # FAISS search
    distances, indices = index.search(query_embedding, top_k)

    # Filter dataframe based on the index from FAISS search
    result_df = df.iloc[indices[0]].copy()
    result_df['distance'] = distances[0]

    return result_df

In [25]:
retrieve("Which game has the highest global sales?", index, games, top_k=5)

Rank   Game_Name Platform  Year       Genre        Publisher  \
0       1  Wii Sports      Wii  2006      Sports         Nintendo   
10     11  Nintendogs       DS  2005  Simulation         Nintendo   
13     14     Wii Fit      Wii  2007      Sports         Nintendo   
138   139      Driver       PS  1999      Action   GT Interactive   
77     78     FIFA 16      PS4  2015      Sports  Electronic Arts   

    North_America_Sales   Europe_Sales   Japan_Sales   Other_Sales  \
0         41.49 Million  29.02 Million  3.77 Million  8.46 Million   
10         9.07 Million   11.0 Million  1.93 Million  2.75 Million   
13         8.94 Million   8.03 Million   3.6 Million  2.15 Million   
138        3.11 Million    2.8 Million  0.02 Million  0.33 Million   
77         1.11 Million   6.06 Million  0.06 Million  1.26 Million   

      Global_Sales                                               Text  \
0    82.74 Million  Rank : 1 | Game_Name : Wii Sports | Platform :...   
10   24.76 Million  Rank : 11 | Game_Name : Nintendogs | Platform ...   
13   22.72 Million  Rank : 14 | Game_Name : Wii Fit | Platform : W...   
138   6.27 Million  Rank : 139 | Game_Name : Driver | Platform : P...   
77    8.49 Million  Rank : 78 | Game_Name : FIFA 16 | Platform : P...   

     distance  
0    0.708280  
10   0.746035  
13   0.753469  
138  0.782200  
77   0.786554

## Applying RAG to LLM

In this section, I implement a **Retrieval-Augmented Generation (RAG)** pipeline using:  
- **LLM Model**: [Gemini 2.0 Flash](https://cloud.google.com/vertex-ai/generative-ai/docs/models/gemini/2-0-flash) (*Free Version*)  
- **Dataset**: [Video Game Sales](https://www.kaggle.com/datasets/gregorut/videogamesales)

The implementation involves creating a custom `generate_answer()` function with the following parameters:   

- `query` : User input question
- `context` : Dataset
- `key` : API key for Gemini authentication

In [26]:
def generate_answer(query, context, key):
    # API Client Setup
    client = OpenAI(
        api_key= key,
        base_url="https://generativelanguage.googleapis.com/v1beta/openai/"
    )
    # Sets the LLM's behavior
    system_message = "You are an intelligent assistant who answers questions based on the provided data."
    # Combines the user's query and retrieved context into a structured prompt
    user_message = f"""
    Question: {query}

    Relevant data:
    {context}
    """
    # Requesting response from LLM
    response = client.chat.completions.create(
        # Gemini model
        model="gemini-2.0-flash",
        messages=[
            {"role": "system", "content": system_message},
            {"role": "user", "content": user_message}
        ]
    )

    return print(response.choices[0].message.content)

### RAG for LLM Tests

In [27]:
# SETUP: Gemini-AI API Key
load_dotenv()
api_key = os.environ['api_key']
api_key

'AIzaSyDuDhaEOZ_eEbQw5ddu-ztvhU-Fe1RoSvM'

In [28]:
generate_answer(query= "What year Wii Sports released?",
                context= games['Text'],
                key= api_key)

Wii Sports released in 2006.


In [29]:
generate_answer(query= "Top 2 Games with highest global sales",
                context= games['Text'],
                key= api_key)

The top 2 games with the highest global sales are:

1.  Wii Sports
2.  Super Mario Bros.


In [30]:
generate_answer(query= "What action game with the highest sales in Europe?",
                context= games['Text'],
                key= api_key)

Based on the provided data:

*   The game "Grand Theft Auto V" on the PS3 platform is the action game with the highest sales in Europe.


In [31]:
generate_answer(query= "How many sales did Grand Theft Auto V on the PS3 had in Europe?",
                context= games['Text'],
                key= api_key)

Based on the provided data:

Grand Theft Auto V on the PS3 had 9.04 million sales in Europe.



In [32]:
generate_answer(query= "List 2 games published by Electronic Arts",
                context= games['Text'],
                key= api_key)

Based on the data provided, I cannot answer the question, as the publisher of the games is not listed.



## Streamlit Application

I’ve deployed this RAG-LLM application as an interactive Streamlit web app. You can test it yourself using the link below. To get started:

1. Prepare a Gemini 2.0 Flash API key

2. Prepare your data in CSV format or use my processed dataset (`games_top200.csv`) included in my [GitHub](https://github.com/andrewanggoh/llm-rag-project)

**Streamlit App**: [aanggoh-llm-rag.streamlit.app](https://aanggoh-llm-rag.streamlit.app)

*Note: Initial loading may take 10-20 seconds (free-tier hosting). Your feedback is welcome!*

### Contacts

- LinkedIn: [@andrewoanggoh](https://www.linkedin.com/in/andrewoanggoh/)
- Gmail: andrew.anggoh@gmail.com